scripts/kconfig/confdata.c

```diff
@@ -636,6 +636,56 @@ static struct conf_printer kconfig_printer_cb =
 	.print_comment = kconfig_print_comment,
 };
 
+/*
+ * rustc cfg printer
+ *
+ * This printer is used when generating the resulting rustc configuration
+ * after kconfig invocation and `defconfig` files.
+ */
+static void rustc_cfg_print_symbol(FILE *fp, struct symbol *sym, const char *value, void *arg)
+{
+	const char *str;
+
+	switch (sym->type) {
+	case S_INT:
+	case S_HEX:
+	case S_BOOLEAN:
+	case S_TRISTATE:
+		str = sym_escape_string_value(value);
+
+		/*
+		 * We don't care about disabled ones, i.e. no need for
+		 * what otherwise are "comments" in other printers.
+		 */
+		if (*value == 'n')
+			return;
+
+		/*
+		 * To have similar functionality to the C macro `IS_ENABLED()`
+		 * we provide an empty `--cfg CONFIG_X` here in both `y`
+		 * and `m` cases.
+		 *
+		 * Then, the common `fprintf()` below will also give us
+		 * a `--cfg CONFIG_X="y"` or `--cfg CONFIG_X="m"`, which can
+		 * be used as the equivalent of `IS_BUILTIN()`/`IS_MODULE()`.
+		 */
+		if (*value == 'y' || *value == 'm')
+			fprintf(fp, "--cfg=%s%s\n", CONFIG_, sym->name);
+
+		break;
+	default:
+		str = value;
+		break;
+	}
+
+	fprintf(fp, "--cfg=%s%s=%s\n", CONFIG_, sym->name, str);
```

Comment:
This creates 2 times the same CFC flag? Does it work? Is cfg!(CONFIG_C) false if X is n? Or is it set to n and so returns true?

Reply:
Yeah, it works! `cfg` holds a multimap, which allows us to do things like this:

    #[cfg(not(CONFIG_X))]
    pub static CONFIG_X_DISABLED: i32 = 0;

    #[cfg(CONFIG_X)]
    pub static CONFIG_X_ENABLED: i32 = 0;
    
    #[cfg(CONFIG_X="y")]
    pub static CONFIG_X_BUILTIN: i32 = 0;
    
    #[cfg(CONFIG_X="m")]
    pub static CONFIG_X_MODULE: i32 = 0;

Since the most common case is checking whether `CONFIG_X` is enabled in any way (either `y` or `m`), I suggested this way in https://github.com/Rust-for-Linux/linux/issues/22#issuecomment-706620374. Thus most code gets the best looking code: `#[cfg(CONFIG_X)]`.

Another option is generating things like `CONFIG_X_ENABLED` directly instead of using just `CONFIG_X`, but I thought this mimicked best what C does, looks best for the common case and is guaranteed to not have collisions, e.g. if someone wants to define both `CONFIG_X` and `CONFIG_X_MODULE`.

Reply:
I should probably copy this example to somewhere in `Documentation/rust` too.